Sources/Basics/Concurrency/ConcurrencyHelpers.swift

```diff
@@ -76,3 +76,55 @@ extension DispatchQueue {
         }
     }
 }
+
+// This implementation is based on the one in swift-build:
+// https://github.com/swiftlang/swift-build/blob/054f2300ad83fd1633f1b50a06b82eea9e7c6901/Sources/SWBUtil/AsyncOperationQueue.swift#L13
+public actor AsyncOperationQueue {
+    private let concurrentTasks: Int
+    private var activeTasks: Int = 0
+    private var waitingTasks: [CheckedContinuation<Void, Never>] = []
+
+    public init(concurrentTasks: Int) {
+        self.concurrentTasks = concurrentTasks
+    }
+
+    deinit {
+        if !waitingTasks.isEmpty {
+            preconditionFailure("Deallocated with waiting tasks")
+        }
+    }
+
+    public func withOperation<ReturnValue>(
+        _ operation: @Sendable () async -> sending ReturnValue
+    ) async -> ReturnValue {
+        await waitIfNeeded()
+        defer { signalCompletion() }
+        return await operation()
+    }
+
+    public func withOperation<ReturnValue>(
+        _ operation: @Sendable () async throws -> sending ReturnValue
+    ) async throws -> ReturnValue {
+        await waitIfNeeded()
+        defer { signalCompletion() }
+        return try await operation()
+    }
+
+    private func waitIfNeeded() async {
+        if activeTasks >= concurrentTasks {
+            await withCheckedContinuation { continuation in
```

Comment:
We should add a `withTaskCancellationHandler` here and potentially a `withTaskPriorityEscalationHandler`

Reply:
No that's not enough since the task can be cancelled after the continuation has been appended. In general, to be structured concurrency and cancellation compliant whenever you create a continuation you need to wrap it into a `withTaskCancellationHandler`. In this case you probably want to remove that continuation from the `waitingTasks` and resume it with a `CancellationError`.

Since you are currently using an `actor` this won't work easily without an unstructured task. So I would recommend you to switch over to a `class` and protect the state with a `Mutex`.

Reply:
Ah, unfortunately SwiftPM's supported `platforms` is set to `.macOS(.v13)` and `Mutex` is only available on >= .v15, so I'll have to use an NSLock as we do elsewhere in SwiftPM.

Reply:
Is there a reason why we can't bump to macOS 15.0?

Reply:
Yes. We cannot require macOS 15 at this time.